Stardrop/Utilities/External/NexusClient.cs

```diff
@@ -349,50 +376,75 @@ public async static Task<bool> ValidateKey(string? apiKey)
             {
                 Program.helper.Log($"Failed to get the download link for Nexus Mods: {ex}", Helper.Status.Alert);
             }
-            client.Dispose();
 
             return null;
         }
 
-        public async static Task<string?> DownloadFileAndGetPath(string uri, string fileName)
+        public async Task<NexusDownloadResult> DownloadFileAndGetPath(string uri, string fileName)
         {
-            // Create a throwaway client
-            HttpClient client = new HttpClient();
-            client.DefaultRequestHeaders.Add("Application-Name", "Stardrop");
-            client.DefaultRequestHeaders.Add("Application-Version", Program.ApplicationVersion);
-            client.DefaultRequestHeaders.Add("User-Agent", $"Stardrop/{Program.ApplicationVersion} {Environment.OSVersion}");
-
+            var requestUri = new Uri(uri);
+            var downloadCancellationSource = new CancellationTokenSource();
+            var requestMessage = new HttpRequestMessage(HttpMethod.Get, requestUri);
             try
             {
-                var stream = await client.GetStreamAsync(new Uri(uri));
-                using (var fileStream = new FileStream(Path.Combine(Pathing.GetNexusPath(), fileName), FileMode.CreateNew))
+
+                var response = await _client.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead, downloadCancellationSource.Token);
+                if (response.IsSuccessStatusCode is false)
                 {
-                    await stream.CopyToAsync(fileStream);
+                    Program.helper.Log($"Failed to download mod file for Nexus Mods: HTTP {response.StatusCode}, {response.ReasonPhrase}", Helper.Status.Alert);
+                    return new(DownloadResultKind.Failed, null);
                 }
 
-                return Path.Combine(Pathing.GetNexusPath(), fileName);
+                using var fileStream = new FileStream(Path.Combine(Pathing.GetNexusPath(), fileName), FileMode.CreateNew);
+                using var downloadStream = await response.Content.ReadAsStreamAsync();
+
+                long? contentLength = response.Content.Headers.ContentLength;
+                DownloadStarted?.Invoke(this, new ModDownloadStartedEventArgs(requestUri, fileName, contentLength, downloadCancellationSource));
+                if (contentLength.HasValue is false || contentLength.Value == 0)
+                {
+                    // We don't know the size, so we can't report progress, so just do a basic downloadStream copy                    
+                    await downloadStream.CopyToAsync(fileStream);
+                }
+                else
+                {
+                    // We *do* know the size, so do manual buffered reads and report progress as we go
+                    var buffer = new byte[81920].AsMemory();
```

Comment:
The `81920` is pretty arbitrary, but is AFAIK, the default HttpClient buffer size.